README.txt
.. image:: https://travis-ci.org/collective/collective.noticeboard.svg?branch=master
    :target: https://travis-ci.org/collective/collective.noticeboard

collective.noticeboard
======================

A fancy noticeboard for Plone inspired by corkboard.me

This Plone-Addon allow suser to transform folders or collections into a fancy noticeboard where users can add and modify postit-like notes.
Notes display the content of a content-type (by default News Items) and are editable in overlays.
Notes are draggable, resizeable and can change color.

collective.noticeboard works with Archetypes and Dexterity. By default it displays

- Title
- Description
- Image (if one exists)
- Author
- Modification-Date

In the configuration you can:

- select the default type used for notes
- select types that should be displayed
- enable adding notes by clicking on the canvas
- enable auto-publishing of notes
- select if notes should be hidden when the were not updated for a certain number of days
- display a link to a help-page (with the id 'noticeboard-help')

Items that were excluded from the navigation are not shown on the noticeboard (you might want to exclude the help-page).

.. warning::

    In Plone, one needs delete permission on an object AND its parent folder
    to delete an item. Most often, a user has the edit permission on
    containing folders, so that does not matter.

    On a noticeboard though, it makes sense that users are only allowed to
    add notes and not to modify the noticeboard itself.

    To allow deletion of content, our delete functionality only checks for the
    delete permission on the object itself and not on the folder.

    This should not create any trouble for you, but we note it here because
    it is a small deviation from Plones default behavior.


Installation
------------

Add this line in the eggs section of your buildout.cfg::

    eggs =
        ...
        collective.noticeboard


Dependencies
------------

- collective.js.jqueryui
- collective.js.backbone
- collective.js.underscore

Take care to choose the right version of collective.js.jqueryui. They offer different versions for each minor Plone Release, so we cannot suggest minimum versions.

To get collective.noticeboard working on Plone 4.3.x you need at least ``collective.js.jqueryui = 1.10.3`` which is not pinned by Plone 4.3.2.

Noticeboard breaks with underscore 1.5.0, so we declared ``collective.js.underscore < 1.5.0`` as a dependency in setup.py.

Editing in overlays should work with the differnt version of Products.TinyMCE shipped with Plone 4.1, 4.2 and 4.3.



Credits
-------

Written by Starzel.de - http://www.starzel.de

Sponsored by DIPF - http://www.dipf.de


Contributors
------------

- Philip Bauer
- Patrick Gerken


CHANGES.txt

0.7.2 (unreleased)
------------------

- Updated buildout using bobotemplates.plone as a template
  [pbauer]

- Enable travis
  [pbauer]


0.7.1 (2014-10-31)
------------------

- Fix security issue, anonymous users could modify notes positions.
  Fixes #6
  [do3cc]


0.7 (2014-10-04)
----------------

- Fix UnicodeError in byline
  [pbauer]

- Set a effective_date on auto-publishing notes
  [pbauer]

- Update to work with newest setuptools
  [pgerken]


0.6 (2013-12-29)
----------------

- Pinn collective.js.underscore to < 1.5.0 to prevent breaking
  [pbauer]

- Prevent empty js-alerts when there is no real error
  [pbauer]

- Try publishing with publish_internally if publish does not work
  [pbauer]

- Add the right js hooks to support dexterity add and edit forms.
  [do3cc]

- Stop using show, but effect. show from jquery and jqueryui have
  different calling args. We actually want to use show from jqueryui
  to have fancy easing effects.
  [do3cc]


0.5 (2013-05-12)
----------------

- fix bug when checking if the item is excluded from navigation
  [pbauer]

- Fix add-url to be compatible with AT and Dexterity
  [pbauer]

- Add some tests
  [do3cc]

- Add compatibility with Plone 4.3
  [pbauer, do3cc]


0.4 (2013-03-06)
----------------

- make buildout use buildout.plonetest
  [toutpt]

- Improve error-message when trying to delete a locked object.
  [pbauer]

- Better Cache Control (prevent anonymous users lock the noticeboard for the next 10 minutes...)
  [pgerken]

- remove obsolte code
  [pbauer]


0.3 (2013-02-05)
----------------

- Fix use with collections
  [pbauer]

- Reimplement Delete functionality. Now deletion of a note is possible
  even without edit rights on the folder. This behavior is different
  to plone stock behavior!
  [do3cc]


0.2 (2013-01-12)
----------------

- add byline
  [pbauer]

- add and improve some colors
  [pbauer]


0.1 (2012-12-11)
----------------

- initial release


src/collective/noticeboard/resources/noticeboard.js
/*jslint nomen: true */
(function init(noticeboard, $, _, Backbone) {
    "use strict"; /*global _: true, jQuery: true, Backbone: true, window: true, Mustache: true, TinyMCEConfig: true, InitializedTinyMCEInstances: true, copyDataForSubmit: true, alert: true */

    Backbone.emulateHTTP = true;

    noticeboard.init = function (canvas) {
        var Note = Backbone.Model.extend({
            url: function () {
                return this.collection.itemurl + '/' + this.id + '/json';
            }

        }),
            Notes = Backbone.Collection.extend({
                initialize: function () {
                    this.on("updateZIndex", this.updateZIndex);
                    this.on("hide_edit", this.hideEdit);
                    this.on("error", this.error);
                },
                model: Note,
                error: function (model, xhr, options) {

                    if (xhr.status === 423) {
                        alert("This note is currently being edited by another user!");
                    } else if (xhr.status === 0) {
                        //don't alert when there is no real error
                    }
                    else {
                        alert("An error occured! Errorcode: " + xhr.status + " Message: " + xhr.statusText);
                    }
                },
                updateZIndex: function () {
                    _.each(_.sortBy(this.models, function (note) {
                        return note.get("zIndex");
                    }), function (note, index) {
                        note.set({
                            zIndex: index
                        });
                        note.save();
                    });
                },
                hideEdit: function () {
                    this.each(function (note) {
                        if (note.get("show_edit")) {
                            note.set({
                                show_edit: 0
                            });
                        }
                    });
                }
            }),

            NoteView = Backbone.View.extend({
                className: "boardnote",
                initialize: function () {
                    this.model.bind("change", this.render, this);
                    this.model.bind("destroy", this.remove, this);
                    _.bindAll(this);
                },
                update: function () {
                    this.model.fetch();
                },
                remove: function () {
                    this.$el.remove();
                },
                updateZIndex: function () {
                    var biggest = _.reduce($(".boardnote"), function (a, b) {
                        return Math.max($(b).zIndex(), a);
                    }, 0);
                    if (this.model.get("zIndex") !== biggest) {
                        this.model.set({
                            zIndex: biggest + 1
                        });
                        // Every once in a while we should decrease zIndex numbers
                        if (Math.random() * 1001 > 1000 || biggest > 1000) {
                            this.model.trigger("updateZIndex");
                        } else {
                            this.model.save();
                        }
                    }

                },
                updateEditBar: function () {
                    this.model.trigger("hide_edit");
                    this.model.set({
                        'show_edit': 1
                    });
                },
                delete1: function (event) {
                    // Two steps. to allow confirmation
                    var model = this.model;
                    this.$el.find(".delete_confirm").effect("slide", {
                        direction: "right"
                    }, function () {
                        var $this = $(this);
                        $this.find(".question, .not_confirm").click(function () {
                            model.trigger("hide_edit");
                            return false;
                        });
                        $this.find(".confirm").click(function () {
                            var href = $(this).attr("href");
                            model.destroy({
                                wait: true
                            });
                        });
                    });
                    event.preventDefault();
                    return false;
                },
                render: function () {
                    var data = {},
                        model = this.model,
                        color = this.model.get("color"),
                        position_x = this.model.get("position_x"),
                        position_y = this.model.get("position_y"),
                        width = this.model.get("width"),
                        height = this.model.get("height"),
                        template = this.model.collection.note_template,
                        publish_link;
                    if (position_x === "25%" && this.model.collection.anon_new_position !== undefined) {
                        this.model.set({
                            "position_x": this.model.collection.anon_new_position.x,
                            "position_y": this.model.collection.anon_new_position.y
                        });
                        this.model.save();
                        return this;
                    }
                    $.extend(data, this.model.toJSON());
                    this.$el.unbind();

                    // A lot of messing with dom is necessary to work properly
                    // with draggable and resizable
                    this.$el.empty();
                    this.$el.removeClass("ui-resizable");
                    this.$el.removeClass("ui-draggable");
                    this.$el.removeData();
                    this.$el.html(Mustache.render(template, data));
                    this.$el.zIndex(this.model.get("zIndex"));
                    if (this.model.get("show_edit")) {
                        this.$el.find(".actions_first").show();
                    }

                    this.$el.css("top", position_y);
                    this.$el.css("left", position_x);
                    this.$el.css("width", width);
                    this.$el.css("height", height);
                    this.$el.css("position", "absolute");
                    if (this.model.get("old_color")) {
                        this.$el.removeClass(this.model.get("old_color"));
                    }
                    this.$el.addClass(color);
                    this.repair_css();
                    this.fix_links();

                    this.$el.draggable({
                        handle: "h3",
                        containment: "window",
                        cursor: "move",
                        stack: ".boardnote",
                        stop: function (object, event) {
                            // We can't prevent draggable from actually
                            // changing the position before we store the
                            // updated position in the model.
                            // So we accept it and update the model without
                            // triggering events that would rerender
                            model.set({
                                position_x: event.position.left,
                                position_y: event.position.top,
                                zIndex: event.helper.zIndex()
                            }, {
                                silent: true
                            });
                            model.save();
                        }
                    });
                    if (this.$el.data('draggable')) {
                        // If there is a data element draggable, we have
                        // to set the position from hand, else movements
                        // somehow mess up the old position
                        // jqueryui 1.10 does not have the data element,
                        // also, it does not behave that way any longer.
                        // 1.10 is the default for Plone 4.3
                        // so expect this clause to get removed when Plone 4.2
                        // is not supported any longer
                        this.$el.data('draggable').position = this.$el.data('draggable').offset = {
                            top: position_y,
                            left: position_y
                        };
                    }
                    this.$el.resizable({
                        minHeight: 150,
                        minWidth: 100,
                        autoHide: true,
                        stop: function (object, event) {
                            // Same tricks as with draggable
                            model.set({
                                width: event.size.width,
                                height: event.size.height
                            }, {
                                silent: true
                            });
                            model.save();
                        }
                    });

                    this.$el.find(".change_color a").click(function (event) {
                        var possible_colors = ['yellow', 'blue', 'green', 'pink', 'purple', 'lightblue', 'grey', 'beige'],
                            color = model.get('color'),
                            color_index = possible_colors.indexOf(color),
                            next_color = possible_colors[(color_index + 1) % (possible_colors.length)];
                        event.preventDefault();
                        model.set({
                            old_color: color
                        });
                        model.set({
                            color: next_color
                        });
                        model.save();
                    });
                    this.$el.find(".deletex a").prepOverlay({
                        subtype: 'ajax',
                        filter: '#content>*',
                        noform: 'close',
                        formselector: 'form#delete_confirmation',
                        afterpost: _.bind(this.remove, this)
                    });
                    this.$el.find(".edit a").prepOverlay({
                        subtype: 'ajax',
                        filter: '#content>*',
                        formselector: 'form[name=edit_form],form#form',
                        noform: 'close',
                        afterpost: _.bind(this.update, this),
                        config: {
                            onLoad: function (event) {
                                event.stopPropagation();
                                var config, tiny;
                                // old tiny, new tiny
                                if (window.TinyMCEConfig) {
                                    config = new TinyMCEConfig("text");
                                    config.init();
                                } else {
                                    tiny = event.currentTarget.getOverlay().find(".mce_editable");
                                    config = tiny.data('mce-config');
                                    tiny.tinymce(config);
                                }
                                // Various required form hacks
                                event.currentTarget.getOverlay().find('.ArchetypesKeywordWidget select').multiSelect();
                                try {
                                    copyDataForSubmit("form-widgets-display_types");
                                } catch (err) {}
                            },
                            onClose: function () {
                                if (window.InitializedTinyMCEInstances) {
                                    delete InitializedTinyMCEInstances.text;
                                }
                            },
                            closeOnClick: false,
                            fixed: true,
                            speed: 'fast'
                        }
                    });
                    publish_link = this.$el.find(".publish");
                    if (this.model.get("review_state") === 'published') {
                        publish_link.hide();
                    } else {
                        this.$el.find(".publish a").click(function (event) {
                            var $this = $(this);
                            event.preventDefault();
                            $.post(this.href, function () {
                                model.set({
                                    review_state: 'published'
                                });
                            });
                        });
                    }
                    this.$el.bind("click.zindex", this.updateZIndex);
                    this.$el.bind("click.edit", this.updateEditBar);
                    this.$el.find(".delete a").click(this.delete1);
                    return this;
                },
                repair_css: function () {
                    // There is no way to implement the right heights in css
                    var note_height = this.$el.innerHeight(),
                        h3_height = this.$el.find("h3").innerHeight();
                    if (h3_height) {
                        this.$el.find(".notecontent").height((note_height - h3_height) / note_height * 100 + "%");
                    }
                },
                fix_links: function () {
                    this.$el.find(".notecontent a").attr("target", "_blank");
                }
            }),
            App = Backbone.View.extend({
                el: canvas,
                events: {
                    "click": "addAnonymous"
                },
                initialize: function () {
                    var notes = this.notes = new Notes(),
                        update = this.update,
                        tiny;
                    this.notes.url = this.$el.data('href');
                    this.notes.itemurl = this.$el.data('hrefitem');
                    this.notes.bind("add", this.addOne, this);
                    this.notes.bind("reset", this.reset, this);
                    this.notes.bind("all", this.render, this);
                    _.bindAll(this);
                    $(".add_note a").prepOverlay({
                        subtype: 'ajax',
                        filter: '#content>*',
                        formselector: 'form[name=edit_form],form#form',
                        noform: 'close',
                        afterpost: this.update,
                        config: {
                            onLoad: function (event) {
                                var config;
                                // several hacks, similar to prepOveray for edit
                                if (window.TinyMCEConfig) {
                                    config = new TinyMCEConfig("text");
                                    config.init();
                                } else {
                                    tiny = event.currentTarget.getOverlay().find(".mce_editable");
                                    config = tiny.data('mce-config');
                                    tiny.tinymce(config);
                                }
                                event.currentTarget.getOverlay().find('.ArchetypesKeywordWidget select').multiSelect();
                                try {
                                    copyDataForSubmit("form-widgets-display_types");
                                } catch (err) {}
                            },
                            onClose: function () {
                                if (window.InitializedTinyMCEInstances) {
                                    delete InitializedTinyMCEInstances.text;
                                }
                            },
                            closeOnClick: false,
                            fixed: true,
                            speed: 'fast'
                        }
                    });
                    $("#notesettings a").prepOverlay({
                        subtype: 'ajax',
                        filter: '#content>*',
                        formselector: 'form',
                        noform: 'reload',
                        config: {
                            onLoad: function (event) {
                                try {
                                    copyDataForSubmit("form-widgets-display_types");
                                } catch (err) {}
                            }
                        }
                    });
                    $("#viewsettings a").prepOverlay({
                        subtype: 'ajax',
                        filter: '#content>*'
                    });
                    $("#notes_archive a").prepOverlay({
                        subtype: 'ajax',
                        filter: '#content>*'
                    });
                    $("#noticeboard-help a").prepOverlay({
                        subtype: 'ajax',
                        filter: '#content>*'
                    });
                    $.get(this.$el.data("notetemplatehref"), function (data) {
                        notes.note_template = data;
                        notes.fetch();
                    });
                },
                addOne: function (note) {
                    var view = new NoteView({
                        model: note
                    });
                    this.$el.append(view.render().el);
                    view.repair_css();
                },
                addAnonymous: function (event) {
                    if (event.target.id !== 'noticeboardcanvas') {
                        return true;
                    }
                    if ($(event.target).data("addanonymous") !== "True") {
                        $(".actions_first:visible").hide();
                        return true;
                    }
                    var add_link = $(".add_note a").attr("href"),
                        update = this.update,
                        pos_x = event.pageX,
                        pos_y = event.pageY,
                        notes = this.notes;
                    notes.anon_new_position = {
                        x: event.pageX,
                        y: event.pageY
                    };
                    $.get(add_link, function (response) {
                        var edit_form = $(response).find("form[name=edit_form]");
                        edit_form.find("input[name=title]").val($(".anonymoustitle").text());
                        $.get(edit_form.attr("action") + "?" + edit_form.serialize(), function () {
                            update();
                        });
                    });
                },
                reset: function () {
                    this.$el.empty();
                    this.notes.each(this.addOne, this);
                },
                update: function () {
                    this.notes.fetch();
                }
            }),
            app = new App();
    };

}(window.noticeboard = window.noticeboard || {}, jQuery, _, Backbone));